### ISSUE TITLE
Complex number handling convention

### ISSUE THREAD

**Issue by Technici4n** [chalk-lab/Mooncake.jl#768]
Hi, I started working on https://github.com/chalk-lab/Mooncake.jl/issues/766, and was having some trouble understanding how to model some complex operations. It turns out that handling complex numbers in the context of reverse-mode AD is apparently quite tricky.

There is this nice write-up on arXiv: [A tutorial on automatic differentiation with complex numbers](https://arxiv.org/abs/2409.06752). I'll give my own summary here. I just read about this today, so I hope I did not make a mistake in the rest. Otherwise feel free to correct me!

The gist is that an $f: \mathbb C \to \mathbb C$ function is identified to a $g: \mathbb R^2 \to \mathbb R^2$ function. There are then two ways to define the adjoint of the differential:

### Option 1
Starting from a tangent $\bar{y}$, write it as $\bar{y} = \bar{y_1} + i\bar{y_2}$. Then find the dual of the differential such that for any $\dot{x_1}, \dot{x_2}$ this holds:
```math
\langle \begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}, Dg[x]\begin{pmatrix}\dot{x_1} \\ \dot{x_2}\end{pmatrix} \rangle = \langle Dg[x]^\ast\begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}, \begin{pmatrix}\dot{x_1} \\ \dot{x_2}\end{pmatrix} \rangle
```

Finally, define:
```math
Df[x]^\ast(\bar{z}) = \left(Dg[x]^\ast\begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}\right)_1 + i\left(Dg[x]^\ast\begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}\right)_2
```

### Option 2
Starting from a tangent $\bar{y}$, write it as $\bar{y} = \bar{y_1} - i\bar{y_2}$. Then write the dual as before.

Finally, define:
```math
Df[x]^\ast(\bar{z}) = \left(Dg[x]^\ast\begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}\right)_1 - i\left(Dg[x]^\ast\begin{pmatrix}\bar{y_1} \\ \bar{y_2}\end{pmatrix}\right)_2
```

### Comparison
In the simple case of $f(z) = \frac 12 z^2$ whose complex derivative is $f'(z) = z$, with option 1 we get $Df[x]^\ast(1+i\cdot0)=x^*$ (so the conjugate of the complex derivative) whereas with option 2 we get $Df[x]^\ast(1+i\cdot0)=x$.

On the flip side for a $\mathbb C \to \mathbb R$ function, option 1 returns the right gradient from the optimization POV.

Another way to see this is from defining the gradient from the JVP as follows:
```math
\langle \nabla f (x), \dot{x} \rangle = D f[x](\dot{x})
```
If we then want the gradient to be equal to the VJP with 1 as the vector:
1. The usual sesquilinear definition of the inner product mandates option 1.
2. A bilinear definition of the inner product mandates option 2.

Related discussion: https://github.com/jax-ml/jax/issues/4891

It seems that JAX implements option 2 whereas TensorFlow and PyTorch implement option 1. I have no idea what was done in other Julia AD packages.

Looks like Mooncake doesn't really support complex numbers atm:
```jl
f(z) = 1/2*z^2
value_and_gradient(f, AutoMooncake(), 1+2im)
```
fails with
```
ERROR: Mooncake.ValueAndGradientReturnTypeError("When calling __value_and_gradient!!, return value of primal must be a subtype of IEEEFloat. Instead, found value of type ComplexF64.")
```
Nonetheless, there might still be $\mathbb R \to \mathbb R$ functions to differentiate that contain complex numbers internally, so to implement the relevant custom rules there should be a clear definition of what an rrule should do for complex numbers.

### Conclusion
I think that Mooncake should make a choice of either option, and nicely document it.

**Comment by Technici4n** [comment]
In terms of Wirtinger derivatives:

Option 1 gives $Df[z]^\ast\bar{z}= \left(\frac{\partial f}{\partial z}\right)^\ast \bar{z} + \frac{\partial f}{\partial z^\ast} \bar{z}^*$

Option 2 gives $Df[z]^\ast\bar{z}= \frac{\partial f}{\partial z} \bar{z} + \frac{\partial f}{\partial z^\ast} \bar{z}^*$

Option 2 thus with $\bar{z}=1$ gives the correct derivative for a holomorphic function.

### Matrix-vector product
As an example, consider the usual matrix-vector product $f(x) = Ax$. The pullback to $A$ is affected by the choice of complex convention:
1. Option 1 gives $\bar{A} = \bar{y} x^H$
2. Option 2 gives $\bar{A} = \bar{y} x^T$

(Assuming that my math is correct 😅)

In this case the extra conjugate from option 1 helps ensure that we get an adjoint rather than a transpose (the latter being a bit unusual for complex arrays).

**Comment by willtebbutt** [comment]
Thanks for writing this up.

I'll take a proper look later today, but I'm fairly sure that we've already implicitly chosen one of the above by
1. letting the tangent type for complex numbers be the default type (some specific parametrisation of `Mooncake.Tangent` -- you can see which one by asking for `Mooncake.tangent_type(ComplexF64)`), and
2. defining `Mooncake._dot` on these types.

That is: we've picked whichever convention is consistent with treating complex numbers like any other `struct` containing two real numbers.

I definitely agree that we need to document this though!

**Comment by Technici4n** [comment]
Hi,

Indeed Mooncake is then using "option 1" since the gradient of $f(z)=z^2/2$ at $z=1+i$ is $1-i$.

I managed to test it with the following code:
```jl
import Mooncake

f(z) = z^2/2

z̄ = 1.0 + im
cache = Mooncake.prepare_pullback_cache(f, z̄; debug_mode=true)
fdata = Mooncake.NoFData()
rdata = Mooncake.RData((; re=1.0, im=0.0))
tangent = Mooncake.tangent(fdata, rdata)
Mooncake.value_and_pullback!!(cache, tangent, f, z̄)
```
Is there no nicer way to construct and deconstruct the tangents? 😅

In any case the convention is clear and I will send a PR adding a relevant docs section.

**Comment by willtebbutt** [comment]
> In any case the convention is clear and I will send a PR adding a relevant docs section.

Fantastic -- I look forward to it!

> Is there no nicer way to construct and deconstruct the tangents? 😅

Haha, yeah, the internals in Mooncake can be a bit verbose. In your example above, it ought to be fine to just construct a `Tangent` directly, rather than making the fdata / rdata, and then combining them. Is this what you're referring to, or is it a different aspect of the above?

**Comment by Technici4n** [comment]
Yes, constructing the Tangent directly should work. It's just that it took me quite a bit of trial and error to get it right. I suppose that it's fine for now, and it was a good exercise!
